zerver/tests/test_presence.py

```diff
@@ -87,6 +87,44 @@ def back_date(num_weeks: int) -> None:
         self.assert_length(presence_dct, 0)
         self.assertEqual(last_update_id, -1)
 
+    def test_user_presence_row_creation_simulated_race(self) -> None:
+        """
+        There is a theoretical race condition, where while a UserPresence
+        row is being created for a user, a concurrent process creates it first,
+        right before we execute our INSERT. This conflict is handled with
+        ON CONFLICT DO NOTHING in the SQL query and an early return
+        if that happens.
+
+        Such a race is difficult to replicate in a test, but the least we can do
+        is simulate the conditions that do_update_user_presence uses to detect
+        that the ON CONFLICT DO NOTHING case happened.
```

Comment:
I would have expected it to work to simulate the race to just mock `cursor.execute` to first create the row separately and then call the actual function. Maybe that's awkward?

Reply:
Hmm yeah I think that's a good idea. I thought it'd be harder to do the mocks or that Django's weird transaction model in tests would break it, but it actually works by setting up a `side_effect` which calls a custom `insert_row_and_return_cursor()` function at the right time. Pushed with that tweak - and I think that's a nice improvement to how real the test is in simulating the race.